lib/Sema/LegalConstExprVerifier.cpp

```diff
@@ -134,13 +134,184 @@ static bool supportedOperator(const ApplyExpr *operatorApplyExpr) {
   return true;
 }
 
+// Per SE-0492:
+// A constant expression is syntactically one of:
+// - an integer literal using any of standard integer types (Int, UInt,
+//   Int8/16/32/64/128, UInt8/16/32/64/128)
+// - a floating-point literal of type Float or Double
+// - a boolean literal of type Bool
+// - a direct reference to a non-generic function using its name (the function
+//   itself is not generic, and also it must not be defined in a generic
+//   context)
+// - a direct reference to a non-generic metatype using the type name directly
+//   (the type itself is not generic, and also it must not be defined in a
+//   generic context), where the type is non-resilient
+// - a tuple composed of only other constant expressions
+// - an array literal of type InlineArray composed of only other constant
+//   expressions
+static std::optional<std::pair<const Expr *, IllegalConstErrorDiagnosis>>
+checkSupportedWithSectionAttribute(const Expr *expr,
+                                   const DeclContext *declContext) {
+  SmallVector<const Expr *, 4> expressionsToCheck;
+  expressionsToCheck.push_back(expr);
+  while (!expressionsToCheck.empty()) {
+    const Expr *expr = expressionsToCheck.pop_back_val();
+
+    // Look through IdentityExpr
+    if (const IdentityExpr *identityExpr = dyn_cast<IdentityExpr>(expr)) {
+      expressionsToCheck.push_back(identityExpr->getSubExpr());
+      continue;
+    }
+    
+    // Tuples composed of constant expressions are allowed
+    if (const TupleExpr *tupleExpr = dyn_cast<TupleExpr>(expr)) {
+      for (const Expr *element : tupleExpr->getElements())
+        expressionsToCheck.push_back(element);
+      continue;
+    }
+
+    // Array literals of type InlineArray composed of constant expressions are
+    // allowed
+    if (const ArrayExpr *arrayExpr = dyn_cast<ArrayExpr>(expr)) {
+      // Check if this is specifically an InlineArray type
+      auto arrayType = arrayExpr->getType();
+      if (arrayType && arrayType->getNominalOrBoundGenericNominal()) {
+        auto nominal = arrayType->getNominalOrBoundGenericNominal();
+        if (nominal->getName().str() == "InlineArray") {
```

Comment:
FYI, you can use `arrayType->isInlineArray()` to check whether it's an `InlineArray`; you don't need to dig out the nominal and check its name as a string.